### PR TITLE
suggest `..` for erroneous `...` in struct pattern; don't discard successfully-parsed fields on error

### PR DESCRIPTION
**Pull request by zackmdavis**
full_field_parse

Regarding #46718.

r? @pnkfelix

### CODE REVIEW

**Comment by xfix** [comment]
I wonder if it wouldn't be better to not print a message about missing field at all here. Using `...` sorta shows the intent to use struct update syntax, where missing fields aren't an issue.

File: src/libsyntax/parse/parser.rs [hunk 1]
```diff
@@ -3426,40 +3426,52 @@ impl<'a> Parser<'a> {
...
             let lo = self.span;
             let hi;
 
             if self.check(&token::DotDot) {
                 self.bump();
                 if self.token != token::CloseDelim(token::Brace) {
                     let token_str = self.this_token_to_string();
-                    return Err(self.fatal(&format!("expected `{}`, found `{}`", "}",
-                                       token_str)))
+                    let err = self.fatal(&format!("expected `{}`, found `{}`", "}",
```
Comment (estebank) [inline src/libsyntax/parse/parser.rs]: `format!("expected `{}`...", "}", ...)` is a bit silly, could you change it? :)

File: src/test/ui/did_you_mean/issue-46718-expected-dotdot-found-dotdotdot.stderr [hunk 2]
```diff
@@ -0,0 +1,14 @@
...
+  --> $DIR/issue-46718-expected-dotdot-found-dotdotdot.rs:21:55
+   |
+21 |             PersonalityInventory { expressivity: exp, ... } => exp
+   |                                                       ^^^ help: to omit remaining fields, use one fewer `.`: `..`
+
+error[E0027]: pattern does not mention field `instrumentality`
+  --> $DIR/issue-46718-expected-dotdot-found-dotdotdot.rs:21:13
+   |
+21 |             PersonalityInventory { expressivity: exp, ... } => exp
+   |             ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^ missing field `instrumentality`
```
Comment (estebank) [inline src/test/ui/did_you_mean/issue-46718-expected-dotdot-found-dotdotdot.stderr]: I agree with @xfix's comment, it'd be nice if this error wasn't shown. IIRC, this error happens during type check, so you'd have to modify the `ast::StructPatKind` and its equivalent in the `hir` to include a boolean `recovered` and only emit this error if it's true (I'm doing the same for type errors involving tail expressions and recovered if blocks). I'm sure that there're are other errors that could benefit from this signal too, but I wouldn't go looking to fix them on this PR. This change would probably cause many more files to be modified, so we could land this first and have a follow up PR if you prefer.

**Review by estebank — APPROVED** [review]
I like it. Agree with @xfix's suggestion but I'd accept a follow up PR for it.

File: src/libsyntax/parse/parser.rs [hunk 3]
```diff
@@ -3426,40 +3426,52 @@ impl<'a> Parser<'a> {
...
                 if self.check(&token::CloseDelim(token::Brace)) { break }
             }
 
-            let attrs = self.parse_outer_attributes()?;
+            let attrs = self.parse_outer_attributes()
+                .map_err(|err| ((fields.clone(), etc), err))?;
             let lo = self.span;
             let hi;
 
             if self.check(&token::DotDot) {
```
Comment (petrochenkov) [inline src/libsyntax/parse/parser.rs]: The only thing this PR needs to do is to replace this `self.check(&token::DotDot)` with `self.check(&token::DotDot) || self.token == token::DotDotDot`, and report a non-fatal error (`self.span_err(...)`) if it's actually a `DotDotDot`.

Comment (zackmdavis) [reply src/libsyntax/parse/parser.rs]: I see ...

(I might claim that the extra machinery in this PR is still potentially useful for taking successfully-parsed field patterns into account when there's some other kind of syntax error ... but maybe that can be a separate, future PR _if_ I find this `PartialPResult` device being useful elsewhere.)

**Comment by zackmdavis** [comment]
Closing in favor of simpler alternative #46763; thanks to @petrochenkov for the suggestion. (The `PartialPResult` machinery here could end up being incorporated in some future proposal, but it's overkill for #46721 alone.)

**Pull request closed without merging**